Compare record counts in Elasticsearch and the Database

atsmith813 opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - Read the DEV Contributing Guide and the Code of Conduct
     - Provided tests for your changes
     - Used descriptive commit messages
     - Updated any relevant documentation and added any necessary screenshots
-->

## What type of PR is this? (check all applicable)
- [x] Feature

## Description
As we migrate more of our search functionality to Elasticsearch, we want to be better about monitoring Elasticsearch. Part of that is to check the number of records in our database compared to the number of records in Elasticsearch to make sure they are in sync. If they're not, it means something is wrong. This PR creates a rake task and a worker to be run daily to perform a check of these counts.

## Related Tickets & Documents
https://github.com/thepracticaldev/dev.to/projects/6#card-34722425

## Added tests?
- [x] yes

## Added to documentation?
- [x] no documentation needed

## Are there any post deployment tasks we need to perform?
1. Someone with the Heroku juice will need to schedule the `log_daily_database_and_elasticsearch_record_counts` task to run daily.
2. Someone with production console juice will need to manually trigger `"elasticsearch"` errors in a production console so I can set up monitors in Datadog.

## [optional] What gif best describes this PR or how it makes you feel?

![lots_of_yellow_rangers_gif](https://media.giphy.com/media/nocBODZsF3KDu/giphy.gif)

atsmith813 commented on app/services/search/base.rb:
```diff
@@ -37,6 +37,10 @@ def update_mappings(index_alias: self::INDEX_ALIAS)
 
+      def db_count
+        raise "Search classes must implement their own db_count"
+      end
```
This isn't used yet. Once a search like `FeedContent` (uses multiple models) is finished, it'll leverage this by adding `db_count` to `Search::FeedContent` with logic to return the appropriate count.

atsmith813 commented on app/workers/metrics/record_db_and_es_record_counts_worker.rb:
```diff
@@ -0,0 +1,64 @@
+
+    # Adjustable margin of error - this is how far off the index count can be
+    # from the database count before we raise an error
+    def perform(margin_of_error = 0)
```
There's a chance we'll get some false errors running this. For example, if we run a query to count the records in the DB and then query ES for records, it's possible another record was created between those queries, making the counts not match. I added this `margin_of_error` argument to make it easy to adjust it in case our notifications get too noisy (change this to 3 or 5 for example).

It's certainly possible we want to extend this further by having `margin_of_error` be index specific. That's okay because we'd likely add that to the search class itself and we wouldn't have to update the arguments of this worker.

atsmith813 commented on app/workers/metrics/record_db_and_es_record_counts_worker.rb:
```diff
@@ -0,0 +1,64 @@
+          record_difference: record_difference,
+          margin_of_error: margin_of_error,
+          action: "record_count"
+        }
```
I figured it can't hurt to send all the data as tags so we can debug very quickly and make some granular monitors in Datadog.

atsmith813 commented on app/workers/metrics/record_db_and_es_record_counts_worker.rb:
```diff
@@ -0,0 +1,64 @@
+          DatadogStatsClient.increment("elasticsearch", tags: tags)
+
+          # This will force the job to retry
+          raise DbAndEsRecordsCountMismatch, "#{search_class} record count mismatch"
```
Basically, I'm intentionally failing the job here in the case of a mismatch so it tries again.

rhymes replied:
I fear this could lead to a never ending queue of these jobs in Sidekiq. I don't remember what's the algorithm for retries but I guess it's exponential, so if the next tick (the next time this job is scheduled) happens this job is going to be scheduled twice, then maybe three times and if it keeps failing because the delta is enough to trigger the change, you get a never ending cascade of jobs.

I also think you sould run 1 job per class. It will also lower the chances that this job gets stuck and you can have them running independently. Basically add the search class name as an argument and pass it from the rake task.

What do you think?

atsmith813 commented on lib/tasks/fetch.rake:
```diff
@@ -128,3 +128,7 @@ end
 end
+
+task log_daily_database_and_elasticsearch_record_counts: :environment do
+  Metrics::RecordDbAndEsRecordCountsWorker.perform_async
```
I could see a case where this namespace isn't the best. While the other workers in `Metrics` actually log using Timber as the task names suggest, the behavior is still very similar and eventually, we'll be sending all logs to Datadog anyways.

citizen428 replied:
If you pick up my name suggestion above, maybe `Search::ReconciliationWorker` or `Search::Statsworker` could be alternatives?

atsmith813 commented on spec/workers/metrics/record_db_and_es_record_counts_worker_spec.rb:
```diff
@@ -0,0 +1,69 @@
+      Search::Tag.refresh_index
+      worker.perform
+
+      tags = { search_class: Search::Tag, db_count: 1, index_count: 1, record_difference: 0, margin_of_error: 0, action: "record_count", record_count: "match" }
```
This is just to make the linter happy. Making it more readable takes up several lines and goes over the limit.

atsmith813 commented on spec/workers/metrics/record_db_and_es_record_counts_worker_spec.rb:
```diff
@@ -0,0 +1,69 @@
+      # Piggyback off of Tags index for testing
+      stub_const "Search::Base::INDEX_ALIAS", Search::Tag::INDEX_ALIAS
+      create(:tag, :search_indexed)
+      Search::Tag.refresh_index
```
I'm open to changing this. I'm just trying to mimic the `tags` index so that there's a valid ES index to get a count from.

atsmith813 commented on app/workers/metrics/record_db_and_es_record_counts_worker.rb:
```diff
@@ -0,0 +1,64 @@
+# need to implement a custom db_count method on the Search class to do the
+# counts for us
+module Metrics
+  class RecordDbAndEsRecordCountsWorker
```
Any suggestions for a shorter name? I was just trying to be super explicit 🙃

atsmith813 commented on app/workers/metrics/record_db_and_es_record_counts_worker.rb:
```diff
@@ -0,0 +1,64 @@
+      return model.count if model.respond_to?(:count)
+
+      search_class.db_count
+    end
```
I'm basically saying the "default" behavior is for a `Search` class to be related to a single model. That's true...for now. So we try to just grab the model count and if we can't we fall back to a custom implementation of the `db_count` method on the `Search` class. In the future, if we start having many more searches that are not tied to a single model, we can change this up to handle it more succinctly.

citizen428 commented:
> Someone with the Heroku juice will need to schedule the log_daily_database_and_elasticsearch_record_counts task to run daily.

What are we using right now, the Heroku scheduler addon? As an alternative I used the [sidekiq-scheduler](https://github.com/moove-it/sidekiq-scheduler) extension in the past, very nice, especially since it means you can make job scheduling part of your PRs.

citizen428 commented on app/workers/metrics/record_db_and_es_record_counts_worker.rb:
```diff
@@ -0,0 +1,64 @@
+    # from the database count before we raise an error
+    def perform(margin_of_error = 0)
+      SEARCH_CLASSES.each do |search_class|
+        margin_of_error = margin_of_error.to_i
```
The JSON data types survive the serialization roundtrip, so you may not need the `to_i`.

> The Sidekiq server pulls that JSON data from Redis and uses JSON.load to convert the data back into Ruby types to pass to your perform method.

That said, should we maybe make this a percentage instead of an absolute number? Models we have more documents of, like articles, probably also have a higher turnover, so maybe also a higher chance for temporary drift? Not sure that's really the best approach, just throwing it out there.

rhymes replied:
It's true that index workers have high priority but there could be a case where the queue is a bit overwhelmed. Maybe the default shouldn't be 0 or as @citizen428 suggsted it should be about percentages instead of relative numbers. Not sure. Maybe we can go ahead with this and if you start getting too many false positives notifications, you can tune this

atsmith813 commented:
@citizen428 I think we are using the Heroku scheduler addon based on the [docs](https://docs.dev.to/backend/scheduled-jobs/).

I like the idea of being able to schedule jobs as part of our PRs! [sidekiq-scheduler](https://github.com/moove-it/sidekiq-scheduler) looks very simple too 💯. Maybe someone with Heroku powers can chime in here with thoughts? I'm also curious about just how many routine jobs we currently have running. Is it just the ones in `/fetch.rake`?

citizen428 commented on app/workers/metrics/record_db_and_es_record_counts_worker.rb:
```diff
@@ -0,0 +1,64 @@
+    def perform(margin_of_error = 0)
+      SEARCH_CLASSES.each do |search_class|
+        margin_of_error = margin_of_error.to_i
+        db_count = db_count(search_class)
```
I know you haven't implemented this yet, but `COUNT()` in Postgres always does a full-table scan, which we may not want to do for several big models in a loop. @rhymes added an `estimated_count` method to `ApplicationRecord`, maybe we should try to use that.

* https://www.citusdata.com/blog/2016/10/12/count-performance/#dup_counts_estimated_full
* https://stackoverflow.com/a/48391562/4186181

atsmith813 replied:
Ooooo this is very interesting - TIL 🤔.

I don't think it's going to work since we need the exact count and we're racing against the clock. I tried `estimated_count` locally too and the counts were off. I could see that difference leading to false failures.

citizen428 replied:
I think this ties to my comment above re using a percentage instead of an absolute value as an indicator for drift.

rhymes replied:
@atsmith813 `.estimated_count` locally is off because we have a "toy DB" and the vacuum cleaner doesn't really work. Same for the stats collector. It's the same reason why it's hard to make accurate estimates about query plans in `development` as our DB has order of magnitudes less data in some tables than production.

I worry `db_count` might end up slowing down the site eventually. It might not be a problem for now as I don't think any of the searchable models are tied to massive tables. According to the API - `curl https://dev.to/api/articles | jq '.[0].id'` we have 286939 published articles and that's probably the biggest table.

I just tried to run the query `ApplicationRecord.estimated_count` runs using Blazer:

```sql
SELECT (
  (reltuples / GREATEST(relpages, 1)) *
  (pg_relation_size('articles') / (GREATEST(current_setting('block_size')::integer, 1)))
)::bigint AS count
FROM pg_class WHERE relname = 'articles';
```

and it returns 0 but I think it's because Blazer doesn't allow me to query that table. 
I can see it works in production since https://dev.to/partnerships displays the estimated count of users.

You might want to verify using the production console to check how off the count is for articles.

If you're confident about what `.estimated_count` please use that plus a percentage as suggested by @citizen428 which should lead to less errors in monitoring

atsmith813 replied:
Ahhhh, I see. That makes sense, thank you! TIL.

I don't have access to a production console so I have no idea. I'm hesitant about percentages 🤔 I'm thinking I might save that for the future.

My thinking right now is to get this monitor in place to help with verifying the new search integrations to help catch bugs like this one https://github.com/thepracticaldev/dev.to/pull/6517 sooner. We caught that because the count was off by 1. Granted, over time, that count would've grown and eventually would've passed some threshold. But I'm thinking of keeping the margin tight while we're implementing new searches and then widening it from there. The `FeedContent` search that Molly is currently working on is the biggest and will be a good test of noise. I'm thinking once we take Algolia out we'll have a sense of sensitivity for this one and can move to percentages.

To your point about Articles likely being the biggest, it's possible we may want to adjust the sensitivity by index too, which would be implemented differently.

We also already have a gap to account for with timing. If we add an estimated count then the gap will become even wider and less accurate.

This task runs once a day with 5 retries. Maybe we schedule it to run in off-peak hours? So that translates to 5 queries max, per index, per day. Would that help it at all?

rhymes commented:
> @citizen428 I think we are using the Heroku scheduler addon based on the [docs](https://docs.dev.to/backend/scheduled-jobs/).
> 
> I like the idea of being able to schedule jobs as part of our PRs! [sidekiq-scheduler](https://github.com/moove-it/sidekiq-scheduler) looks very simple too 💯. Maybe someone with Heroku powers can chime in here with thoughts? I'm also curious about just how many routine jobs we currently have running. Is it just the ones in `/fetch.rake`?

IIRC @mstruve planned to eventually move off Heroku scheduler to have everything inside Sidekiq, but I think she wanted to wait until ES is up and running everywhere on the site and everything is running smoothly

rhymes commented on app/services/search/base.rb:
```diff
@@ -37,6 +37,10 @@ def update_mappings(index_alias: self::INDEX_ALIAS)
       end
 
+      def db_count
+        raise "Search classes must implement their own db_count"
```
+1 on a custom Exception. You can add it to `app/errors` if you want

rhymes commented on app/workers/metrics/record_db_and_es_record_counts_worker.rb:
```diff
@@ -0,0 +1,64 @@
+    private
+
+    def db_count(search_class)
+      model = search_class.to_s.split("::").last.safe_constantize
```
`"Search::ChatChannelMembership".demodulize.safe_constantize` it's a trick I picked up from @citizen428

atsmith813 replied:
It's not a string at this point so you end up getting `NoMethodError: undefined method 'demodulize' for Search::ChatChannelMembership:Class`. Do you mean using `.to_s.demodulize` instead of `split("::").last`?

atsmith813 commented:
@rhymes 

> I fear this could lead to a never ending queue of these jobs in Sidekiq. I don't remember what's the algorithm for retries but I guess it's exponential, so if the next tick (the next time this job is scheduled) happens this job is going to be scheduled twice, then maybe three times and if it keeps failing because the delta is enough to trigger the change, you get a never ending cascade of jobs.

I don't think so. This would still be treated as a normal failing job, right? Similar to if there was a syntax error or `nil` error of some sort? Sidekiq would see the job errored out, and queue a retry up to the limit (5). Then the job would be moved to the morgue and marked as dead if it failed all 5 times. I was just explicitly raising an error so Sidekiq didn't treat the job as successful. Does that make sense?

> I also think you sould run 1 job per class. It will also lower the chances that this job gets stuck and you can have them running independently. Basically add the search class name as an argument and pass it from the rake task.

I love this idea! It will be better too because that way, whenever we update the list of indexes we're counting, there aren't old jobs hanging with the old list. It's always fresh. Easier to debug too that way if one index fails in the job, that job doesn't retry going through successful indexes again.

rhymes commented:
> I don't think so. This would still be treated as a normal failing job, right? Similar to if there was a syntax error or `nil` error of some sort? Sidekiq would see the job errored out, and queue a retry up to the limit (5). Then the job would be moved to the morgue and marked as dead if it failed all 5 times. I was just explicitly raising an error so Sidekiq didn't treat the job as successful. Does that make sense?

I think it depends on how often that scheduling works. Most jobs are not scheduled so they fail and repeat N times until they either succeed or die off. But if between the second to last and the last repeat cycle the next iteration of the scheduler runs, then you have two jobs that might fail in the queue. Just a theory, but anyway, thanks for your explanation here https://github.com/thepracticaldev/dev.to/pull/6801#discussion_r397322699

>  I love this idea! It will be better too because that way, whenever we update the list of indexes we're counting, there aren't old jobs hanging with the old list. It's always fresh. Easier to debug too that way if one index fails in the job, that job doesn't retry going through successful indexes again.

Yeah, also there's no reason why those jobs can't be run in parallel as they all operate on different tables.

atsmith813 commented:
@rhymes @citizen428 made a handful of updates. Summary below.

- Rename the worker to `Search::ReconciliationWorker`.
- Update the job to work on a single `search_class` - moving the loop to the rake task.
- Added functionality to handle percentages for `margin_of_error`.
- Added functionality to use `estimated_count`.
- Updated tags being sent to Datadog.

I added optional arguments to the job `margin_of_error` and `use_estimated_count`.
- `margin_of_error` this handles both percentages and record counts. Probably not the best overall for clarity sake, but bear with me 🙃. If `margin_of_error` is a value greater than 0 and less than 1, it will be treated as a percent. For example, if `margin_of_error` is `0.10`, it will be treated as 10%. If the value of `margin_of_error` is greater than 1, equal to 1, or equal to 0, the value will be treated as a number of records. For example, a value of `2` would treat this as 2 records. If the difference in counts is more than 2, an error is raised.
- `use_estimated_count` this defaults to false. If it's set to true, inside the `db_count` helper method it will try to use `estimated_count` first when calculating `db_count`. Something to note is that some models (i.e. `Tag`) don't inherit from `ApplicationRecord` so that method (`estimated_count`) is not available. It can also be `nil` in testing. So when `use_estimated_count` is true, it will try to use `estimated_count` first, then `Model.count`, then the custom `Search::Base.db_count` method to determine a `db_count`.

I added `percentage_difference` to tags so we can track that. I really would like to understand just how severe the gap in counts is just from timing alone. So my thinking is to leave the `margin_of_error` and `use_estimated_count` to `0` and `false` respectively. Then I can monitor just how big the gaps are both in absolute terms and percentage terms (via all the info being sent over as tags). From there we can make more informed decisions on what to set these parameters and it only requires a quick change to the rake the arguments being passed into the rake task and no change to the job itself.

Once I understand the gap from timing, I plan to adjust the `margin_of_error` to account for that and then set `use_estimated_count` to true. Then we can more accurately understand what the gap is from using `estimated_counts` and adjust the `margin_of_error` more accurately from there.

It's worth noting this likely won't happen for some time. The existing indexes and models we are testing here are already fully in sync (previously tested in a prod console by Molly), so I don't suspect these monitors to even really trigger until we add new searches to it.

mstruve commented on app/workers/search/reconciliation_worker.rb:
```diff
@@ -0,0 +1,75 @@
+      search_class.db_count
+    end
+
+    def count_mismatched?(margin_of_error, percentage_difference, record_difference)
```
You know what, let's keep this simple and just log the counts in Datadog then we can use Datadog alerts to determine what is a "problematic" threshold and what is not.

atsmith813 commented:
@mstruve updated. Let me know if I misunderstood something.

atsmith813 commented:
This PR is officially a dumpster fire. Throwing it away and starting over 🚮.
